relay-system/src/controller.rs

```diff
@@ -40,190 +71,205 @@ impl ShutdownHandle {
-    /// actors have completed a graceful shutdown.
-    pub fn run<F, R, E>(shutdown_timeout: Duration, factory: F) -> Result<(), E>
-    where
-        F: FnOnce() -> Result<R, E>,
-    {
-        // Spawn a legacy actix system for the controller's signals.
-        let sys = actix::System::new("relay");
-
-        // Ensure that the controller starts if no service has starts it. It will register with
-        // `ProcessSignals` shut down even if no actors have subscribed. If we remove this line, the
-        // controller will not be instantiated and our system will not listen for signals.
-        CONTROLLER
-            .set(Controller::new(shutdown_timeout).start())
-            .ok();
-
-        // Run the factory and exit early if an error happens. The return value of the factory is
-        // discarded for convenience, to allow shorthand notations.
-        factory()?;
-
-        // All actors have started successfully. Run the system, which blocks the current thread
-        // until a signal arrives or `Controller::stop` is called.
-        sys.run();
-
-        Ok(())
+    /// Initiates the shutdown process of the system.
+    pub fn trigger_shutdown(mode: ShutdownMode) {
+        let (ref tx, _) = *MANUAL_SHUTDOWN;
+        tx.send(Some(mode)).ok();
     }
 
     /// Returns a [handle](ShutdownHandle) to receive shutdown notifications.
     pub fn shutdown_handle() -> ShutdownHandle {
-        let (_, ref rx) = SHUTDOWN.get_or_init(|| watch::channel(None));
+        let (_, ref rx) = *SHUTDOWN;
         ShutdownHandle(rx.clone())
     }
 
-    /// Performs a graceful shutdown with the given timeout.
+    /// Wait for the shutdown and timeout.
     ///
-    /// This sends a `Shutdown` message to all subscribed actors and waits for them to finish. As
-    /// soon as all actors have completed, `Controller::stop` is called.
-    fn shutdown(&mut self, context: &mut Context<Self>, timeout: Option<Duration>) {
-        // Send a shutdown signal to all registered subscribers (including self). They will report
-        // when the shutdown has completed. Note that we ignore all errors to make sure that we
-        // don't cancel the shutdown of other services if one service fails.
-        let (tx, _) = SHUTDOWN.get_or_init(|| watch::channel(None));
-        tx.send(Some(Shutdown { timeout })).ok();
-
-        // Delay the shutdown for 100ms to allow recipients of the shutdown signal to execute their
-        // error handlers. Once `System::stop` is called, futures won't be polled anymore and we
-        // will not be able to print error messages.
-        let when = timeout.unwrap_or_else(|| Duration::from_secs(0)) + Duration::from_millis(100);
+    /// This waits for the first shutdown signal and then conditionally waits for the shutdown
+    /// timeout. If the shutdown timeout is interrupted by another signal, this function resolves
+    /// immediately.
+    pub async fn shutdown() {
```

Comment:
The duality of `shutdown` and `shutdown_handle` is not ideal. I'm now thinking to refactor one more time so that:

- `Controller` has `shutdown_handle()` and `trigger_shutdown`. 
- `ShutdownHandle` has `async notified()` and `async finished()`.